src/localplayer.cpp

```diff
@@ -319,46 +235,43 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
 	bool touching_ground_was = touching_ground;
 	touching_ground = result.touching_ground;
 
+	// Max. distance (X, Z) over border for sneaking determined by collision box
+	// * 0.49 to keep the center just barely on the node
+	v3f sneak_max = m_collisionbox.getExtent() * 0.49;
+
 	// We want the top of the sneak node to be below the players feet
 	f32 position_y_mod;
 	if (m_sneak_node_exists)
 		position_y_mod = m_sneak_node_bb_top.MaxEdge.Y - 0.05 * BS;
 	else
 		position_y_mod = (0.5 - 0.05) * BS;
-	v3s16 current_node = floatToInt(position - v3f(0, position_y_mod, 0), BS);
-	/*
-		Check the nodes under the player to see from which node the
-		player is sneaking from, if any.  If the node from under
-		the player has been removed, the player falls.
-	*/
-	if (m_sneak_node_exists &&
-			nodemgr->get(map->getNodeNoEx(m_old_node_below)).name == "air" &&
-			m_old_node_below_type != "air") {
-		// Old node appears to have been removed; that is,
-		// it wasn't air before but now it is
-		m_need_to_get_new_sneak_node = false;
+
+	// Get position of current standing node
+	const v3s16 current_node = floatToInt(position - v3f(0, position_y_mod, 0), BS);
+
+	if (current_node != m_sneak_node) {
 		m_sneak_node_exists = false;
-	} else if (nodemgr->get(map->getNodeNoEx(current_node)).name != "air") {
-		// We are on something, so make sure to recalculate the sneak
-		// node.
-		m_need_to_get_new_sneak_node = true;
+	} else {
+		node = map->getNodeNoEx(current_node, &is_valid_position);
+		if (!is_valid_position || !nodemgr->get(node).walkable)
+			m_sneak_node_exists = false;
 	}
 
-	if (m_need_to_get_new_sneak_node && physics_override_sneak) {
-		v2f player_p2df(position.X, position.Z);
+	if (could_sneak && !m_sneak_node_exists) {
+		// Get new sneak node
 		f32 min_distance_f = 100000.0 * BS;
-		v3s16 new_sneak_node = m_sneak_node;
-		for(s16 x=-1; x<=1; x++)
-		for(s16 z=-1; z<=1; z++)
-		{
-			v3s16 p = current_node + v3s16(x,0,z);
-			v3f pf = intToFloat(p, BS);
-			v2f node_p2df(pf.X, pf.Z);
-			f32 distance_f = player_p2df.getDistanceFrom(node_p2df);
+		m_sneak_ladder_detected = false;
+
+		for (s16 x = -1; x <= 1; x++)
+		for (s16 z = -1; z <= 1; z++) {
+			const v3s16 p = current_node + v3s16(x,0,z);
+			const v3f pf = intToFloat(p, BS);
+			const v2f diff(position.X - pf.X, position.Z - pf.Z);
+			f32 distance_f = diff.getLength();
 
 			if (distance_f > min_distance_f ||
-					fabs(player_p2df.X-node_p2df.X) > (.5+.1)*BS + sneak_max.X ||
-					fabs(player_p2df.Y-node_p2df.Y) > (.5+.1)*BS + sneak_max.Z)
+					fabs(diff.X) > 0.6 * BS + sneak_max.X ||
```

Comment:
Where does this magic number come from at all? I would like to document this properly.